aws-greengrass/aws-greengrass-nucleus
# Unsubscribe on shutdown

**Issue #, if available:**
1. Subscription timeouts happen when nucleus starts while there are already jobs waiting in the queue. 
2. subscriptions are persisted across nucleus restarts due to mqtt client using same session. On shutdown the subscription handlers are killed but the client has subscriptions when the nucleus restarts

**Description of changes:**
1. Changed the order of subscriptions so that there are handlers for job description before notification subscription is created. This solves the timeout issue.
2. Unsubscribing from Iot Jobs topics in deployment service shutdown. This should keep the session clean.

**Why is this change necessary:**

**How was this change tested:**
Manually ran GGC using the jar. Created cloud deployments and ran following scenarios:
1. start nucleus, create deployment (supposed to succeed), deployment succeeded, restart nucleus manually. Did not see deployment replaying
2. start nucleus, create deployment (supposed to fail), deployment failed, restart nucleus manually. Did not see deployment replaying
3. create a deployment (supposed to fail), start nucleus, deployment failed, create another deployment (supposed to fail), deployment failed.

Observed no timeouts and no replaying of deployments in above scenarios with these changes. 

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
